File: app/src/main/java/com/termux/app/managers/ScriptManager.java
package com.termux.app.managers;

import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;
import com.termux.app.models.ScriptItem;
import com.termux.shared.logger.Logger;

import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class ScriptManager {
    private static final String TAG = "ScriptManager";
    private static final String SCRIPTS_DIR = "scripts";
    private static final String SCRIPTS_INDEX = "scripts.json";

    private static ScriptManager instance;
    private List<ScriptItem> cachedScripts;

    public static ScriptManager getInstance() {
        if (instance == null) {
            instance = new ScriptManager();
        }
        return instance;
    }

    public List<ScriptItem> getScripts(Context context) {
        if (cachedScripts != null) {
            return cachedScripts;
        }

        cachedScripts = new ArrayList<>();

        try {
            String indexContent = readAssetFile(context, SCRIPTS_DIR + "/" + SCRIPTS_INDEX);
            ScriptsIndex index = new Gson().fromJson(indexContent, ScriptsIndex.class);

            if (index != null && index.scripts != null) {
                for (ScriptEntry entry : index.scripts) {
                    try {
                        String content = readAssetFile(context, SCRIPTS_DIR + "/" + entry.file);
                        cachedScripts.add(new ScriptItem(entry.name, entry.description, entry.file, content));
                    } catch (Exception e) {
                        Logger.logWarn(TAG, "Failed to load script: " + entry.file);
                    }
                }
            }
        } catch (Exception e) {
            Logger.logError(TAG, "Failed to load scripts index: " + e.getMessage());
        }

        return cachedScripts;
    }

    private String readAssetFile(Context context, String path) throws IOException {
        try (Reader reader = new InputStreamReader(context.getAssets().open(path), StandardCharsets.UTF_8)) {
            StringBuilder sb = new StringBuilder();
            char[] buffer = new char[1024];
            int len;
            while ((len = reader.read(buffer)) != -1) {
                sb.append(buffer, 0, len);
            }
            return sb.toString();
        }
    }

    public void clearCache() {
        cachedScripts = null;
    }

    static class ScriptsIndex {
        @SerializedName("scripts")
        List<ScriptEntry> scripts;
    }

    static class ScriptEntry {
        String name;
        String description;
        String file;
    }
}

File: app/src/main/java/com/termux/app/managers/ScriptManager.kt
package com.termux.app.managers

import android.content.Context
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.termux.app.models.ScriptItem
import com.termux.shared.logger.Logger
import java.io.IOException

class ScriptManager private constructor() {

    companion object {
        private const val LOG_TAG = "ScriptManager"
        private const val SCRIPTS_DIR = "scripts"
        private const val SCRIPTS_INDEX = "scripts.json"

        @Volatile
        private var instance: ScriptManager? = null

        @JvmStatic
        fun getInstance(): ScriptManager {
            return instance ?: synchronized(this) {
                instance ?: ScriptManager().also { instance = it }
            }
        }
    }

    private var cachedScripts: List<ScriptItem>? = null

    data class ScriptsIndex(
        val scripts: List<ScriptEntry>
    )

    data class ScriptEntry(
        val name: String,
        val description: String,
        val file: String
    )

    fun loadScripts(context: Context): List<ScriptItem> {
        cachedScripts?.let { return it }

        val scripts = mutableListOf<ScriptItem>()

        try {
            // Read scripts.json index
            val indexContent = context.assets.open("$SCRIPTS_DIR/$SCRIPTS_INDEX").bufferedReader().use { it.readText() }
            val index = Gson().fromJson(indexContent, ScriptsIndex::class.java)

            // Load each script
            for (entry in index.scripts) {
                try {
                    val content = context.assets.open("$SCRIPTS_DIR/${entry.file}").bufferedReader().use { it.readText() }
                    scripts.add(ScriptItem(
                        name = entry.name,
                        description = entry.description,
                        fileName = entry.file,
                        content = content
                    ))
                } catch (e: IOException) {
                    Logger.logWarn(LOG_TAG, "Failed to load script: ${entry.file}")
                }
            }
        } catch (e: IOException) {
            Logger.logError(LOG_TAG, "Failed to load scripts index: ${e.message}")
        }

        cachedScripts = scripts
        return scripts
    }

    fun getScripts(context: Context): List<ScriptItem> {
        return cachedScripts ?: loadScripts(context)
    }

    fun getScriptContent(context: Context, name: String): String? {
        return getScripts(context).find { it.name == name }?.content
    }

    fun clearCache() {
        cachedScripts = null
    }
}